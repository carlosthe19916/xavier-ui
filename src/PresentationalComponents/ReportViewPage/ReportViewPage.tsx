import React, { Fragment, Component } from 'react';
import { Redirect } from 'react-router-dom';
import {
    Main,
    PageHeader,
    PageHeaderTitle,
    Skeleton
} from '@redhat-cloud-services/frontend-components';
import {
    Tabs,
    Tab,
    Breadcrumb,
    BreadcrumbItem
} from '@patternfly/react-core';
import { Report } from '../../models';
import { Link } from 'react-router-dom';
import { RouterGlobalProps } from '../../models/router';
import { REPORT_VIEW_PATHS } from '../../pages/ReportView/ReportViewConstants';
import { ObjectFetchStatus } from '../../models/state';

export interface Props extends RouterGlobalProps {
    mainStyle?: any;
    report: Report | null;
    reportFetchStatus: ObjectFetchStatus;
};

interface State {
    activeTabKey: number
}

class ReportViewPage extends Component<Props, State> {

    constructor(props: Props) {
        super(props);

        let activeTabKey = 1;

        if (props.location.pathname.endsWith(REPORT_VIEW_PATHS.workloadMigrationSummary)) {
            activeTabKey = 0;
        } else if (props.location.pathname.endsWith(REPORT_VIEW_PATHS.workloadMigrationSummary)) {
            activeTabKey = 1;
        } else if (props.location.pathname.endsWith(REPORT_VIEW_PATHS.workloadInventory)) {
            activeTabKey = 2;
        }

        this.state = {
            activeTabKey
        };
    }

    public handleTabClick = (_event: any, tabIndex: number) => {
        this.setState({
            activeTabKey: tabIndex
        });

        const { history, match } = this.props;

        switch (tabIndex) {
            case 0:
                history.push(`${match.url}/${REPORT_VIEW_PATHS.workloadMigrationSummary}`);
                break;
            case 1:
                history.push(`${match.url}/${REPORT_VIEW_PATHS.initialSavingsEstimation}`);
                break;
            case 2:
                history.push(`${match.url}/${REPORT_VIEW_PATHS.workloadInventory}`);
                break;
        }
    };

    public renderTabs = () => {
        const { report } = this.props;
        const currentBreadcrumb = report ? report.reportName : '';

        return (
            <React.Fragment>
                <Breadcrumb style={ { marginBottom: '25px' } }>
                    <BreadcrumbItem>
                        <Link to="/reports">Reports</Link>
                    </BreadcrumbItem>
                    <BreadcrumbItem isActive={true}>{ currentBreadcrumb }</BreadcrumbItem>
                </Breadcrumb>
                <Tabs
                    isFilled={true}
                    onSelect={ this.handleTabClick }
                    activeKey={ this.state.activeTabKey }
                >
                    <Tab eventKey={ 0 } title="Workload Migration Summary"/>
                    <Tab eventKey={ 1 } title="Initials Savings Estimation"/>
                    <Tab eventKey={ 2 } title="Workload Inventory"/>
                </Tabs>
            </React.Fragment>
        );
    }

    public renderTabsSkeleton = () => {
        return (
            <React.Fragment>
                <div className="pf-l-stack pf-m-gutter">
                    <div className="pf-l-stack__item">
                        <Skeleton size="sm" />
                    </div>
                    <div className="pf-l-stack__item">
                        <div className="pf-l-grid">
                            <div className="pf-l-grid__item pf-m-4-col">
                                <Skeleton size="md" />
                            </div>
                            <div className="pf-l-grid__item pf-m-4-col">
                                <Skeleton size="md" />
                            </div>
                            <div className="pf-l-grid__item pf-m-4-col">
                                <Skeleton size="md" />
                            </div>
                        </div>
                    </div>
                </div>
            </React.Fragment>
        );
    };

    public render() {
        const { reportFetchStatus, children } = this.props;

        if (reportFetchStatus.error) {
            return <Redirect to={ `/reports` } />;
        }

        const isFetchComplete: boolean = reportFetchStatus.status === 'complete';

        return (
            <Fragment>
                <PageHeader>
                    <PageHeaderTitle title={ isFetchComplete ? this.renderTabs() : this.renderTabsSkeleton() } />
                </PageHeader>
                <Main style={ this.props.mainStyle }>
                    { isFetchComplete ? children : '' }
                </Main>
            </Fragment>
        );
    }
}

export default ReportViewPage;
