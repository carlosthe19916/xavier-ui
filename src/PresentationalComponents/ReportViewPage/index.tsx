import ReportViewPage from './ReportViewPage';
import { withRouter } from 'react-router';

export default withRouter(ReportViewPage);
