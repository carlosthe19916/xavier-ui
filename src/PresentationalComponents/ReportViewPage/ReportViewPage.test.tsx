import React from "react";
import { shallow } from "enzyme";
import ReportViewPage, { Props } from "./ReportViewPage";

const baseProps: Props = {
  report: {
    id: 123,
    creationDate: 123,
    customerId: '123',
    fileName: '123.zip',
    numberOfHosts: 123,
    totalDiskSpace: 123,
    totalPrice: 123,
    status: 'IN_PROGRESS'
  },
  reportFetchStatus: {
    error: null,
    status: "complete"
  },

  match: null,
  history: null,
  location: {
    pathname: ''
  }
};

describe("ReportViewPage", () => {

  it("expect to redirect on fetchReport error", () => {
    const props: Props = {
      ...baseProps,
      reportFetchStatus: {
        ...baseProps.reportFetchStatus,
        error: 'A error message'
      }
    };

    const wrapper = shallow(
      <ReportViewPage mainStyle={ { color: 'red' } } { ...props } >
        children
      </ReportViewPage>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("expect to render skeleton", () => {
    const props: Props = {
      ...baseProps,
      reportFetchStatus: {
        ...baseProps.reportFetchStatus,
        error: null,
        status: "inProgress"
      }
    };

    const wrapper = shallow(
      <ReportViewPage mainStyle={ { color: 'red' } } { ...props } >
        children
      </ReportViewPage>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("expect to render tabs when fetch status is complete", () => {
    const props: Props = {
      ...baseProps,
      reportFetchStatus: {
        ...baseProps.reportFetchStatus,
        error: null,
        status: "complete"
      }
    };

    const wrapper = shallow(
      <ReportViewPage mainStyle={ { color: 'red' } } { ...props } >
        children
      </ReportViewPage>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("expect to render tabs when fetch status is complete and report is null", () => {
    const props: Props = {
      ...baseProps,
      report: null,
      reportFetchStatus: {
        ...baseProps.reportFetchStatus,
        error: null,
        status: "complete"
      }
    };

    const wrapper = shallow(
      <ReportViewPage mainStyle={ { color: 'red' } } { ...props } >
        children
      </ReportViewPage>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
