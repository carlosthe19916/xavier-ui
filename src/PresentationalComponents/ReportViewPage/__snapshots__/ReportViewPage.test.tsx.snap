// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportViewPage expect to redirect on fetchReport error 1`] = `
<Redirect
  push={false}
  to="/reports"
/>
`;

exports[`ReportViewPage expect to render skeleton 1`] = `
<Fragment>
  <PageHeader>
    <PageHeaderTitle
      title={
        <React.Fragment>
          <div
            className="pf-l-stack pf-m-gutter"
          >
            <div
              className="pf-l-stack__item"
            >
              <Skeleton
                size="sm"
              />
            </div>
            <div
              className="pf-l-stack__item"
            >
              <div
                className="pf-l-grid"
              >
                <div
                  className="pf-l-grid__item pf-m-4-col"
                >
                  <Skeleton
                    size="md"
                  />
                </div>
                <div
                  className="pf-l-grid__item pf-m-4-col"
                >
                  <Skeleton
                    size="md"
                  />
                </div>
                <div
                  className="pf-l-grid__item pf-m-4-col"
                >
                  <Skeleton
                    size="md"
                  />
                </div>
              </div>
            </div>
          </div>
        </React.Fragment>
      }
    />
  </PageHeader>
  <Connect(Main)
    style={
      Object {
        "color": "red",
      }
    }
  />
</Fragment>
`;

exports[`ReportViewPage expect to render tabs when fetch status is complete 1`] = `
<Fragment>
  <PageHeader>
    <PageHeaderTitle
      title={
        <React.Fragment>
          <Breadcrumb
            style={
              Object {
                "marginBottom": "25px",
              }
            }
          >
            <BreadcrumbItem>
              <Link
                replace={false}
                to="/reports"
              >
                Reports
              </Link>
            </BreadcrumbItem>
            <BreadcrumbItem
              isActive={true}
            >
              My report name 123
            </BreadcrumbItem>
          </Breadcrumb>
          <Tabs
            activeKey={1}
            className=""
            isFilled={true}
            isSecondary={false}
            leftScrollAriaLabel="Scroll left"
            onSelect={[Function]}
            rightScrollAriaLabel="Scroll Right"
            variant="div"
          >
            <ForwardRef
              eventKey={0}
              title="Workload Migration Summary"
            />
            <ForwardRef
              eventKey={1}
              title="Initials Savings Estimation"
            />
            <ForwardRef
              eventKey={2}
              title="Workload Inventory"
            />
          </Tabs>
        </React.Fragment>
      }
    />
  </PageHeader>
  <Connect(Main)
    style={
      Object {
        "color": "red",
      }
    }
  >
    children
  </Connect(Main)>
</Fragment>
`;

exports[`ReportViewPage expect to render tabs when fetch status is complete and report is null 1`] = `
<Fragment>
  <PageHeader>
    <PageHeaderTitle
      title={
        <React.Fragment>
          <Breadcrumb
            style={
              Object {
                "marginBottom": "25px",
              }
            }
          >
            <BreadcrumbItem>
              <Link
                replace={false}
                to="/reports"
              >
                Reports
              </Link>
            </BreadcrumbItem>
            <BreadcrumbItem
              isActive={true}
            >
              
            </BreadcrumbItem>
          </Breadcrumb>
          <Tabs
            activeKey={1}
            className=""
            isFilled={true}
            isSecondary={false}
            leftScrollAriaLabel="Scroll left"
            onSelect={[Function]}
            rightScrollAriaLabel="Scroll Right"
            variant="div"
          >
            <ForwardRef
              eventKey={0}
              title="Workload Migration Summary"
            />
            <ForwardRef
              eventKey={1}
              title="Initials Savings Estimation"
            />
            <ForwardRef
              eventKey={2}
              title="Workload Inventory"
            />
          </Tabs>
        </React.Fragment>
      }
    />
  </PageHeader>
  <Connect(Main)
    style={
      Object {
        "color": "red",
      }
    }
  >
    children
  </Connect(Main)>
</Fragment>
`;
